### PR TITLE
Changing the following classes and define the properties as immutable SemanticVersion, SemanticVersionPreReleaseTag and SemanticVersionBuildMetaData

### LINKED ISSUE #3443: [Improvement] The following classes needs to be changed to immutable types: SemanticVersion, SemanticVersionPreReleaseTag and SemanticVersionBuildMetaData
**Is your improvement request related to a problem? Please describe.**
The following classes nees to be change to immutable types: SemanticVersion, SemanticVersionPreReleaseTag and SemanticVersionBuildMetaData

### PR DESCRIPTION
**Pull request by HHobeck**
Changing the following classes and define the properties as immutable SemanticVersion, SemanticVersionPreReleaseTag and SemanticVersionBuildMetaData

Close #3443 

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/GitVersion.Core/VersionCalculation/VariableProvider.cs [hunk 1]
```diff
@@ -2,64 +2,80 @@
...
         if (isContinuousDeploymentMode)
         {
             semanticVersion = new SemanticVersion(semanticVersion);
             // Continuous Deployment always requires a pre-release tag unless the commit is tagged
             if (!semanticVersion.PreReleaseTag.HasTag())
             {
-                var label = configuration.GetBranchSpecificLabel(semanticVersion.BuildMetaData.Branch, null);
-                semanticVersion.PreReleaseTag.Name = label ?? string.Empty;
-                if (semanticVersion.PreReleaseTag.Name.IsNullOrEmpty())
+                // TODO: Why do we manipulating the semantic version here in the VariableProvider? The method name is GET not MANIPULATE.
```
Comment (arturcic) [inline src/GitVersion.Core/VersionCalculation/VariableProvider.cs:28]: I wonder if these TODOs are still relevant taking in consideration we compute the semantic version and return it. We could rename the method to describe the 'Compute' logic

File: src/GitVersion.Core/VersionCalculation/VariableProvider.cs [hunk 2]
```diff
@@ -2,64 +2,80 @@
...
             {
-                var match = Regex.Match(semanticVersion.BuildMetaData.Branch, configuration.LabelNumberPattern);
-                var numberGroup = match.Groups["number"];
-                if (numberGroup.Success)
-                {
-                    // TODO: Why do we manipulating the semantic version here in the VariableProvider? The method name is GET not MANIPULATE.
-                    // What is about the separation of concern and single-responsibility principle?
-                    semanticVersion.PreReleaseTag.Name += numberGroup.Value;
-                }
+                // TODO: Why do we manipulating the semantic version here in the VariableProvider? The method name is GET not MANIPULATE.
```
Comment (arturcic) [inline src/GitVersion.Core/VersionCalculation/VariableProvider.cs:46]: I wonder if these TODOs are still relevant taking in consideration we compute the semantic version and return it

**Review by arturcic — APPROVED** [review]

File: src/GitVersion.Core/VersionCalculation/VariableProvider.cs (hunk 2 above)
Comment (HHobeck) [reply src/GitVersion.Core/VersionCalculation/VariableProvider.cs]: Okay no problem I have removed the comment.

File: src/GitVersion.Core/VersionCalculation/VariableProvider.cs (hunk 1 above)
Comment (HHobeck) [reply src/GitVersion.Core/VersionCalculation/VariableProvider.cs]: Okay no problem I have removed the comment.

**Pull request merged**